vmm/src/vm.rs

```diff
@@ -798,76 +826,85 @@ impl Vm {
 
     #[cfg(feature = "fw_cfg")]
     fn populate_fw_cfg(
+        fw_cfg_config: FwCfgConfig,
         device_manager: &Arc<Mutex<DeviceManager>>,
         config: &Arc<Mutex<VmConfig>>,
     ) -> Result<()> {
-        device_manager
-            .lock()
-            .unwrap()
-            .fw_cfg()
-            .expect("fw_cfg device must be present")
-            .lock()
-            .unwrap()
-            .add_e820(config.lock().unwrap().memory.size as usize)
-            .map_err(Error::CreatingE820Map)?;
-        let kernel = config
-            .lock()
-            .unwrap()
-            .payload
-            .as_ref()
-            .map(|p| p.kernel.as_ref().map(File::open))
-            .unwrap_or_default()
-            .transpose()
-            .map_err(Error::MissingFwCfgKernelFile)?;
-        if let Some(kernel_file) = kernel {
-            device_manager
-                .lock()
-                .unwrap()
-                .fw_cfg()
-                .expect("fw_cfg device must be present")
-                .lock()
-                .unwrap()
-                .add_kernel_data(&kernel_file)
-                .map_err(Error::MissingFwCfgKernelFile)?
+        let mut e820_option: Option<usize> = None;
+        if fw_cfg_config.e820 {
+            e820_option = Some(config.lock().unwrap().memory.size as usize);
         }
-        let cmdline = Vm::generate_cmdline(
-            config.lock().unwrap().payload.as_ref().unwrap(),
-            #[cfg(target_arch = "aarch64")]
-            device_manager,
-        )
-        .map_err(|_| Error::MissingFwCfgCmdline)?
-        .as_cstring()
-        .map_err(|_| Error::MissingFwCfgCmdline)?;
-        device_manager
-            .lock()
-            .unwrap()
-            .fw_cfg()
-            .expect("fw_cfg device must be present")
-            .lock()
-            .unwrap()
-            .add_kernel_cmdline(cmdline);
-        let initramfs = config
-            .lock()
-            .unwrap()
-            .payload
-            .as_ref()
-            .map(|p| p.initramfs.as_ref().map(File::open))
-            .unwrap_or_default()
-            .transpose()
-            .map_err(Error::MissingFwCfgInitramfs)?;
-        // We measure the initramfs when running Oak Containers in SNP mode (initramfs = Stage1)
-        // o/w use Stage0 to launch cloud disk images
-        if let Some(initramfs_file) = initramfs {
-            device_manager
+        let mut kernel_option: Option<File> = None;
+        if fw_cfg_config.kernel {
+            let kernel = config
                 .lock()
                 .unwrap()
-                .fw_cfg()
-                .expect("fw_cfg device must be present")
+                .payload
+                .as_ref()
+                .map(|p| p.kernel.as_ref().map(File::open))
+                .unwrap_or_default()
+                .transpose()
+                .map_err(Error::MissingFwCfgKernelFile)?;
```

Comment:
It should be handled in a central place for the validation of CLI options for the combination of payload. See my comments from #7166. 

I just want to call it out here and let's leave it for future cleanup once #7166 is landed when @phip1611 is back.